### ISSUE TITLE
putting getindex(::NTuple, ::NTuple) into standard library

### ISSUE THREAD

**Issue by mchristianl** [JuliaLang/julia#47870]
Hi,

It might be a bit of a question of taste, but ...
Would it be possible to add a `getindex` method to the standard library such that tuples can be indexed with tuples?

E.g.

```julia
t = (10,20,30,40)
σ = (2,4)
t[σ] # MethodError: no method matching getindex(::NTuple{4, Int64}, ::Tuple{Int64, Int64}) ↯
ntuple(i -> t[σ[i]],length(σ)) :: NTuple # (20,40) works as expected ✓
```

This kind of operation seems to occur a lot when computing with **permutations**: e.g. having a tuple of the 4 corner points of a tetrahedron and then iterating over all its edges. This can be done in such a way

```julia
@noinline function edge_length(Ps::NTuple{4,NTuple{3}})
  l = 0.0
  @inline norm(v) = sqrt(max(0,sum(v .* v))) # allow inlining
  σs = ((1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)) # some permutation
  for σ in σs
    Qs = ntuple(i -> Ps[σ[i]],length(σ)) # Qs = Ps[σ]
    l += norm(.-(Qs...)) # do something with Qs
  end
  return l
end

Ps = ((0,0,0),(0,0,1),(0,1,0),(1,0,0))
@code_native debuginfo=:none edge_length(Ps)
using BenchmarkTools
@benchmark edge_length($Ps) # ~50 cycles
```

This might also bother other people, because non-allocating idioms for `Ps[σ]` are
- `ntuple(i -> Ps[σ[i]],length(σ))`
- `Tuple(SVector(Ps)[SVector(σ)])`

so one might use SVector all the time, but these need an `SVector(...)` constructor and several conversions when dealing with a datastructure that uses tuples.

Could there be any reason against this?

kind regards

**Comment by kalmarek** [comment]
maybe it's a bit orthogonal to the issue of adding the syntax but here you're treating a tuple as permutation. If you stop doing it, namely write:
```julia
struct Perm{N,T<:Integer}
    cycle::NTuple{N,T}
end
Base.getindex(t::Tuple, p::Perm{N}) where N = ntuple(i->t[p.cycle[i]], N)

@noinline function edge_length2(Ps::NTuple{4,NTuple{3}}, σs)
  l = 0.0
  @inline norm(v) = sqrt(max(0,sum(v .* v))) # allow inlining
  for σ in σs
    Qs = Ps[σ]
    l += norm(.-(Qs...)) # do something with Qs
  end
  return l
end

σs2 = Perm{2,Int}.(((1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4))) # some permutation
@btime edge_length2($Ps, $σs2)
```
the result is the same and equally fast (at least on my computer: ~12-14 ns)

**Comment by mchristianl** [comment]
Yes, of course. The question is more, whether this calls for an extra type `Perm` ?

You know, there are permutations, combinations, choices, multichoices, inverse permutations, Lehmer code [and whatnot](https://en.wikiversity.org/wiki/Inversion_(discrete_mathematics)). A whole Julia package of stuff. And when you work on these things (inserting, deleting or replacing an index), then you end up with something that is nothing of those. So at the end it is just a tuple of numbers and not more. This seems not worth introducing so many new types (and conversions to interface with it) because it really is this one feature of being able to index a tuple with a tuple, I guess. The name of the new type would be `IndexTuple`.

**Comment by kalmarek** [comment]
heh, I'd say that new types are cheap -- come dozen a penny ;) Especially that in the slack thread you said that the list of those perms is going to be static (and generated?) Adding those 3 lines is much cheaper than adding a new syntax to base julia

> You know, there are permutations, combinations, choices, multichoices, inverse permutations, Lehmer code [and whatnot](https://en.wikiversity.org/wiki/Inversion_(discrete_mathematics)).

That's precisely why I think You do want different types for those -- not only for dispatch, but also to minimize your mental load (in the future) -- so that You don't have to think what does this "tuple" stand for in your function. But anyway -- we're digressing here from the proposal ;)

**Comment by mchristianl** [comment]
Yes, thank you. I had an elaborate discussion on Slack and I think that the example might not be a best fit to show why this would be a nice-to-have. Should we remove the example from the proposal? It might be easier to just get some impression whether there are people who find this syntax helpful or not.

**Comment by mchristianl** [comment]
@mikmoore I have added these to the "idioms" list in the proposal

**Comment by vtjnash** [comment]
Normally we expect array-like functionality to be part of `StaticArrays`, not `NTuple`

**Comment by mchristianl** [comment]
An interesting application would be to have this functionality on an inhomogeneous mixed-type `Tuple` which would not be covered by homogeneous single-type StaticArrays. So just for arguing in favor of the proposal, one might say that multi-index access in the way it is proposed, has an application that is not just array-like.

**Comment by mikmoore** [comment]
Inhomogeneous tuples can already be handled by the suggested `map` or `broadcast` idioms.

The argument against this are largely status-quo arguments (not to discount them). This isn't faster nor significantly shorter to write than the potential versions under the idioms section and it doesn't expand `getindex`'s normal behavior to a previously unsupported index type (it's not like this previously worked for `Array` the proposal extends that support to `Tuple`).

Highly generic functions like `getindex` are useful because they can operate interchangeably over a wide variety of types. Adding special support for special types risks balkanizing `getindex` and eroding its usefulness. Rather, it might be more useful if this were applicable to many container types (somebody would ask for this for `Vector`, at least). However, this interoperability goal would mean committing to `getindex(container,::Tuple)` having these semantics for any `typeof(container)` where this could be sensible. That seems like a big step so shouldn't be taken lightly. For example I would foresee potential conflicts with `Dict` because you'd need to distinguish between tuples *as keys* and tuples *of keys* (although perhaps these could be resolved or we wouldn't care -- `Dict` is already somewhat limited in the ways it can apply `getindex`).

I'd also be anxious about introducing a tripping hazard next to splatting 
```julia
arr = [11 12 13; 21 22 23; 31 32 33]
ind = (2,1)
arr[ind...] # return 21, although I would have suggested the use of `CartesianIndex` for this
arr[ind] # return (21,11) if this proposal were extended to Array, else error
```
where one might have intended either of the indexing expressions. Perhaps the last expression erroring isn't such a bad thing.

I think I'd be more amenable to this as applied to "multidimensional" indexing, e.g, `(11,12,13)`[1,2] == (11,12)` (one could splat a tuple to achieve the proposal here), since multidimensional indexing is very limited in `Base`. But I haven't thought through the full implications of that and am still not sure I'd personally be in favor.

**Comment by mchristianl** [comment]
Thank you for that detailed command! Indeed, I also see that the _consequences_ of such an addition are quite far-reaching when trying to provide a consistent experience with the Julia containers. In that context, I also see that we might save more people from unintended behavior when there is a limit with what's possible with indexing tuples.

So I am totally on your side that this addition, with the implicated additions that follow, needs more thinking and some more time. Your argument, that the last example should fail intentionally _is_ a good argument.

The other side of the coin might be, that performant, fixed-size operations are available only with one of the idioms from the list, or with SVector, or `@SVec`. So, to play devil's advocate, one might say that the performant bits need a special handling ... 

*duck and run
